File: examples/todo_app/swift_todo/Sources/TodoBackend/TodoBridge.swift
// SPDX-License-Identifier: Apache-2.0
// IPC dispatch + JSON (port of c_todo/src/todo_bridge.c; Foundation JSON).

import Foundation
import SQLite3

public typealias TodoBridgeEmit = (_ event: String, _ json: String) -> Void

private func emitWireError(_ emit: TodoBridgeEmit, _ code: Int, _ message: String) {
  let o: [String: Any] = ["code": code, "message": message]
  guard let data = try? JSONSerialization.data(withJSONObject: o),
    let s = String(data: data, encoding: .utf8)
  else { return }
  emit("error", s)
}

private func jStr(_ pl: [String: Any]?, _ key: String) -> String {
  guard let pl, let v = pl[key] else { return "" }
  if v is NSNull { return "" }
  if let s = v as? String { return s }
  if let n = v as? NSNumber { return n.stringValue }
  return ""
}

private func jStrOr(_ pl: [String: Any]?, _ key: String, default def: String) -> String {
  let t = jStr(pl, key)
  return t.isEmpty ? def : t
}

private func jInt64(_ pl: [String: Any]?, _ key: String) -> Int64 {
  guard let pl, let v = pl[key], !(v is NSNull) else { return 0 }
  if let i = v as? Int64 { return i }
  if let i = v as? Int { return Int64(i) }
  if let n = v as? NSNumber { return n.int64Value }
  return 0
}

private func jsonString(data: Any) -> String? {
  guard JSONSerialization.isValidJSONObject(data),
    let d = try? JSONSerialization.data(withJSONObject: data)
  else { return nil }
  return String(data: d, encoding: .utf8)
}

private func categoryRowJson(_ r: TodoStoreCategoryRow) -> [String: Any] {
  ["id": r.id, "name": r.name, "color": r.color]
}

private func todoRowJson(_ r: TodoStoreTodoRow) -> [String: Any] {
  var o: [String: Any] = [
    "id": r.id,
    "title": r.title,
    "description": r.description,
    "status": r.status,
    "priority": r.priority,
    "position": r.position,
    "createdAt": r.createdAt,
    "updatedAt": r.updatedAt,
  ]
  if r.categoryIdIsNull { o["categoryId"] = NSNull() }
  else { o["categoryId"] = r.categoryId }
  if r.parentIdIsNull { o["parentId"] = NSNull() }
  else { o["parentId"] = r.parentId }
  if r.dueDateIsNull { o["dueDate"] = NSNull() }
  else { o["dueDate"] = r.dueDate }
  return o
}

private func emitCategoriesList(_ s: TodoStore, _ emit: TodoBridgeEmit) -> Int {
  var rows: [TodoStoreCategoryRow] = []
  let rc = s.categoriesForeach { rows.append($0) }
  guard rc == SQLITE_OK else {
    emitWireError(emit, Int(rc), "categories.foreach failed")
    return -1
  }
  let arr = rows.map { categoryRowJson($0) }
  guard let json = jsonString(data: arr) else { return -1 }
  emit("categories.list", json)
  return 0
}

private func emitTodosList(_ s: TodoStore, _ emit: TodoBridgeEmit) -> Int {
  var rows: [TodoStoreTodoRow] = []
  let rc = s.todosForeach { rows.append($0) }
  guard rc == SQLITE_OK else {
    emitWireError(emit, Int(rc), "todos.foreach failed")
    return -1
  }
  let arr = rows.map { todoRowJson($0) }
  guard let json = jsonString(data: arr) else { return -1 }
  emit("todos.list", json)
  return 0
}

/// Handle one `todo` wire JSON message; emits follow-up events via `emit`.
public func bridgeHandleWire(_ store: TodoStore, rawIpcJson: String, emit: TodoBridgeEmit) -> Int {
  guard !rawIpcJson.isEmpty else { return -1 }
  guard let rootData = rawIpcJson.data(using: .utf8),
    let rootObj = try? JSONSerialization.jsonObject(with: rootData) as? [String: Any]
  else {
    emitWireError(emit, -1, "invalid wire json")
    return -1
  }
  guard let d = rootObj["d"] as? [String: Any] else {
    emitWireError(emit, -1, "missing wire.d object")
    return -1
  }
  let action = jStr(d, "action")
  if action.isEmpty {
    emitWireError(emit, -1, "missing action")
    return -1
  }
  let pl = d["payload"] as? [String: Any]

  if action == "categories.list" {
    return emitCategoriesList(store, emit)
  }
  if action == "todos.list" {
    return emitTodosList(store, emit)
  }

  if action == "categories.create" {
    guard let pl else {
      emitWireError(emit, Int(SQLITE_MISUSE), "missing payload")
      return -1
    }
    let name = jStr(pl, "name")
    let color = jStrOr(pl, "color", default: "#6366f1")
    var newId: Int64 = 0
    let rc = store.categoryInsert(name: name, color: color, outId: &newId)
    guard rc == SQLITE_OK else {
      emitWireError(emit, Int(rc), "category insert failed")
      return -1
    }
    return emitCategoriesList(store, emit)
  }

  if action == "categories.delete" {
    guard let pl else {
      emitWireError(emit, Int(SQLITE_MISUSE), "missing payload")
      return -1
    }
    let id = jInt64(pl, "id")
    let rc = store.categoryDeleteById(id)
    guard rc == SQLITE_OK else {
      emitWireError(emit, Int(rc), "category delete failed")
      return -1
    }
    return emitCategoriesList(store, emit)
  }

  if action == "todos.create" {
    guard let pl else {
      emitWireError(emit, Int(SQLITE_MISUSE), "missing payload")
      return -1
    }
    let title = jStr(pl, "title")
    let desc = jStrOr(pl, "description", default: "")
    let status = jStrOr(pl, "status", default: "pending")
    let pri = jStrOr(pl, "priority", default: "medium")
    let cat = jInt64(pl, "categoryId")
    let par = jInt64(pl, "parentId")
    let due = jInt64(pl, "dueDate")
    let catSet = cat > 0
    let parSet = par > 0
    let dueSet = due != 0
    var newId: Int64 = 0
    let rc = store.todoInsert(
      title: title,
      description: desc,
      status: status,
      priority: pri,
      categoryIdSet: catSet,
      categoryId: cat,
      parentIdSet: parSet,
      parentId: par,
      dueDateSet: dueSet,
      dueDate: due,
      outId: &newId
    )
    guard rc == SQLITE_OK else {
      emitWireError(emit, Int(rc), "todo insert failed")
      return -1
    }
    return emitTodosList(store, emit)
  }

  if action == "todos.update" {
    guard let pl else {
      emitWireError(emit, Int(SQLITE_MISUSE), "missing payload")
      return -1
    }
    let id = jInt64(pl, "id")
    let title = jStr(pl, "title")
    let desc = jStrOr(pl, "description", default: "")
    let status = jStrOr(pl, "status", default: "pending")
    let pri = jStrOr(pl, "priority", default: "medium")
    let cat = jInt64(pl, "categoryId")
    let par = jInt64(pl, "parentId")
    let due = jInt64(pl, "dueDate")
    let pos = Int(jInt64(pl, "position"))
    let catSet = cat > 0
    let parSet = par > 0
    let dueSet = due != 0
    let rc = store.todoUpdateFull(
      id: id,
      title: title,
      description: desc,
      status: status,
      priority: pri,
      categoryIdSet: catSet,
      categoryId: cat,
      parentIdSet: parSet,
      parentId: par,
      dueDateSet: dueSet,
      dueDate: due,
      position: pos
    )
    guard rc == SQLITE_OK else {
      emitWireError(emit, Int(rc), "todo update failed")
      return -1
    }
    return emitTodosList(store, emit)
  }

  if action == "todos.delete" {
    guard let pl else {
      emitWireError(emit, Int(SQLITE_MISUSE), "missing payload")
      return -1
    }
    let id = jInt64(pl, "id")
    let rc = store.todoDeleteById(id)
    guard rc == SQLITE_OK else {
      emitWireError(emit, Int(rc), "todo delete failed")
      return -1
    }
    return emitTodosList(store, emit)
  }

  emitWireError(emit, -1, "unknown action")
  return -1
}

File: examples/todo_app/csharp_todo/TodoBridge.cs
// IPC dispatch + JSON (port of c_todo/src/todo_bridge.c / nim_todo/todo_bridge.nim).
// SPDX-License-Identifier: Apache-2.0

using System.Text.Json;

namespace Jamharah.Todo;

public static class TodoBridge
{
    private static void EmitErr(Action<string, string> emit, int code, string msg)
    {
        var o = JsonSerializer.Serialize(new { code, message = msg });
        emit("error", o);
    }

    private static string JStr(JsonElement? pl, string k)
    {
        if (pl is not { } root || root.ValueKind != JsonValueKind.Object)
            return "";
        if (!root.TryGetProperty(k, out var v) || v.ValueKind != JsonValueKind.String)
            return "";
        return v.GetString() ?? "";
    }

    private static string JStrOr(JsonElement? pl, string k, string defaultValue)
    {
        var s = JStr(pl, k);
        return s.Length > 0 ? s : defaultValue;
    }

    private static long JInt64(JsonElement? pl, string k)
    {
        if (pl is not { } root || root.ValueKind != JsonValueKind.Object)
            return 0;
        if (!root.TryGetProperty(k, out var v))
            return 0;
        return v.ValueKind switch
        {
            JsonValueKind.Number => v.TryGetInt64(out var n) ? n : 0,
            JsonValueKind.True => 1,
            JsonValueKind.False => 0,
            _ => 0,
        };
    }

    private static object CategoryRowJson(TodoStoreCategoryRow r) =>
        new { id = r.Id, name = r.Name, color = r.Color };

    private static object TodoRowJson(TodoStoreTodoRow r)
    {
        return new
        {
            id = r.Id,
            title = r.Title,
            description = r.Description,
            status = r.Status,
            priority = r.Priority,
            categoryId = r.CategoryIdIsNull ? null : (object?)r.CategoryId,
            parentId = r.ParentIdIsNull ? null : (object?)r.ParentId,
            dueDate = r.DueDateIsNull ? null : (object?)r.DueDate,
            position = r.Position,
            createdAt = r.CreatedAt,
            updatedAt = r.UpdatedAt,
        };
    }

    private static int EmitCategoriesList(TodoStore? s, Action<string, string> emit)
    {
        var rows = new List<TodoStoreCategoryRow>();
        var rc = TodoStoreApi.StoreCategoriesForeach(s, rows.Add);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "categories.foreach failed");
            return -1;
        }
        emit("categories.list", JsonSerializer.Serialize(rows.Select(CategoryRowJson)));
        return 0;
    }

    private static int EmitTodosList(TodoStore? s, Action<string, string> emit)
    {
        var rows = new List<TodoStoreTodoRow>();
        var rc = TodoStoreApi.StoreTodosForeach(s, rows.Add);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "todos.foreach failed");
            return -1;
        }
        emit("todos.list", JsonSerializer.Serialize(rows.Select(TodoRowJson)));
        return 0;
    }

    public static int BridgeHandleWire(TodoStore? s, string rawIpcJson, Action<string, string> emit)
    {
        if (s is null || string.IsNullOrEmpty(rawIpcJson))
            return -1;

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(rawIpcJson);
        }
        catch (JsonException)
        {
            EmitErr(emit, -1, "invalid wire json");
            return -1;
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                EmitErr(emit, -1, "invalid wire json");
                return -1;
            }

            if (!root.TryGetProperty("d", out var d) || d.ValueKind != JsonValueKind.Object)
            {
                EmitErr(emit, -1, "missing wire.d object");
                return -1;
            }

            var action = d.TryGetProperty("action", out var actEl) && actEl.ValueKind == JsonValueKind.String
                ? actEl.GetString() ?? ""
                : "";
            if (action.Length == 0)
            {
                EmitErr(emit, -1, "missing action");
                return -1;
            }

            JsonElement? pld = d.TryGetProperty("payload", out var pl) && pl.ValueKind == JsonValueKind.Object
                ? pl
                : null;

            return action switch
            {
                "categories.list" => EmitCategoriesList(s, emit),
                "todos.list" => EmitTodosList(s, emit),
                "categories.create" => HandleCategoryCreate(s, pld, emit),
                "categories.delete" => HandleCategoryDelete(s, pld, emit),
                "todos.create" => HandleTodoCreate(s, pld, emit),
                "todos.update" => HandleTodoUpdate(s, pld, emit),
                "todos.delete" => HandleTodoDelete(s, pld, emit),
                _ => UnknownAction(emit),
            };
        }
    }

    private static int UnknownAction(Action<string, string> emit)
    {
        EmitErr(emit, -1, "unknown action");
        return -1;
    }

    private static int HandleCategoryCreate(TodoStore? s, JsonElement? pld, Action<string, string> emit)
    {
        if (pld is null)
        {
            EmitErr(emit, SqliteRc.Misuse, "missing payload");
            return -1;
        }
        var name = JStr(pld, "name");
        var color = JStrOr(pld, "color", "#6366f1");
        var rc = TodoStoreApi.StoreCategoryInsert(s, name, color, out _);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "category insert failed");
            return -1;
        }
        return EmitCategoriesList(s, emit);
    }

    private static int HandleCategoryDelete(TodoStore? s, JsonElement? pld, Action<string, string> emit)
    {
        if (pld is null)
        {
            EmitErr(emit, SqliteRc.Misuse, "missing payload");
            return -1;
        }
        var id = JInt64(pld, "id");
        var rc = TodoStoreApi.StoreCategoryDeleteById(s, id);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "category delete failed");
            return -1;
        }
        return EmitCategoriesList(s, emit);
    }

    private static int HandleTodoCreate(TodoStore? s, JsonElement? pld, Action<string, string> emit)
    {
        if (pld is null)
        {
            EmitErr(emit, SqliteRc.Misuse, "missing payload");
            return -1;
        }
        var title = JStr(pld, "title");
        var desc = JStrOr(pld, "description", "");
        var status = JStrOr(pld, "status", "pending");
        var pri = JStrOr(pld, "priority", "medium");
        var cat = JInt64(pld, "categoryId");
        var par = JInt64(pld, "parentId");
        var due = JInt64(pld, "dueDate");
        var rc = TodoStoreApi.StoreTodoInsert(
            s, title, desc, status, pri, cat > 0, cat, par > 0, par, due != 0, due, out _);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "todo insert failed");
            return -1;
        }
        return EmitTodosList(s, emit);
    }

    private static int HandleTodoUpdate(TodoStore? s, JsonElement? pld, Action<string, string> emit)
    {
        if (pld is null)
        {
            EmitErr(emit, SqliteRc.Misuse, "missing payload");
            return -1;
        }
        var id = JInt64(pld, "id");
        var title = JStr(pld, "title");
        var desc = JStrOr(pld, "description", "");
        var status = JStrOr(pld, "status", "pending");
        var pri = JStrOr(pld, "priority", "medium");
        var cat = JInt64(pld, "categoryId");
        var par = JInt64(pld, "parentId");
        var due = JInt64(pld, "dueDate");
        var pos = (int)JInt64(pld, "position");
        var rc = TodoStoreApi.StoreTodoUpdateFull(
            s, id, title, desc, status, pri, cat > 0, cat, par > 0, par, due != 0, due, pos);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "todo update failed");
            return -1;
        }
        return EmitTodosList(s, emit);
    }

    private static int HandleTodoDelete(TodoStore? s, JsonElement? pld, Action<string, string> emit)
    {
        if (pld is null)
        {
            EmitErr(emit, SqliteRc.Misuse, "missing payload");
            return -1;
        }
        var id = JInt64(pld, "id");
        var rc = TodoStoreApi.StoreTodoDeleteById(s, id);
        if (rc != SqliteRc.Ok)
        {
            EmitErr(emit, rc, "todo delete failed");
            return -1;
        }
        return EmitTodosList(s, emit);
    }
}
